File: OC/内容层动画/WGOCCoreAnimation/WGOCCoreAnimation/CAShapeLayer动画/WGBarChartView.m
//
//  WGBarChartView.m
//  WGOCCoreAnimation
//
//  Created by wanggang on 2019/5/27.
//  Copyright © 2019 bozhong. All rights reserved.
//

#import "WGBarChartView.h"

@interface WGBarChartView()

@property (nonatomic, strong) CAShapeLayer *shapeLayer;

@end

@implementation WGBarChartView

@synthesize lineColor = _lineColor;

-(instancetype)initWithFrame:(CGRect)frame{
    
    self = [super initWithFrame:frame];
    if (self) {
        [self createView];
    }
    return self;
}

- (void)createView{
    
    self.backgroundColor = UIColor.whiteColor;
    
    self.shapeLayer = [[CAShapeLayer alloc] init];
    _shapeLayer.fillColor = UIColor.whiteColor.CGColor;
    _shapeLayer.lineWidth = 20;
    _shapeLayer.strokeEnd = 0;
    [self.layer addSublayer:_shapeLayer];
}

-(void)setPointArr:(NSArray *)pointArr{
    
    _pointArr = pointArr;
    
    UIBezierPath *path = [UIBezierPath bezierPath];
    for (NSArray *arr in pointArr) {
        
        for (NSValue *pointValue in arr) {
            
            CGPoint point = [pointValue CGPointValue];
            
            if ([pointValue isEqualToValue:arr[0]]) {
                
                [path moveToPoint:point];
            }else{
                [path addLineToPoint:point];
            }
        }
    }
    
    _shapeLayer.path = path.CGPath;
    _shapeLayer.strokeColor = self.lineColor.CGColor;
}

- (void)startAni{
    
    CABasicAnimation *ani = [[CABasicAnimation alloc] init];
    ani.keyPath = @"strokeEnd";
    ani.fromValue = @(0);
    ani.toValue = @(1);
    [ani setRemovedOnCompletion:NO];
    ani.duration = 3;
    
    _shapeLayer.strokeEnd = 1;
    [_shapeLayer addAnimation:ani forKey:nil];
}

-(UIColor *)lineColor{
    if (!_lineColor) {
        _lineColor = [UIColor redColor];
    }
    return _lineColor;
}

-(void)setLineColor:(UIColor *)lineColor{
    _lineColor = lineColor;
    _shapeLayer.strokeColor = _lineColor.CGColor;
}

@end

File: Swift/内容层动画/WGCoreAnimation/WGCoreAnimation/CAShapeLayer动画/WGBarChartView.swift
//
//  WGBarChartView.swift
//  WGCoreAnimation
//
//  Created by wanggang on 2019/5/8.
//  Copyright © 2019 wanggang. All rights reserved.
//

import UIKit

class WGBarChartView: UIView {

    let barChart = CAShapeLayer()
    let ani = CABasicAnimation()
    
    override init(frame: CGRect) {
        
        super.init(frame: frame)
        
        self.backgroundColor = UIColor.white
        barChart.lineCap = .square
        barChart.lineJoin = .round
        barChart.strokeEnd = 0.0
        barChart.lineWidth = 30
        barChart.fillColor = UIColor.white.cgColor
        self.layer.addSublayer(barChart)
    }
    
    override func draw(_ rect: CGRect) {
        
        let bezierPath = UIBezierPath()
        bezierPath.lineWidth = 30
        bezierPath.lineCapStyle = .square
        bezierPath.lineJoinStyle = .round
        
        for i in 0...4 {
            
            let x = CGFloat(60+70*i)
            let y = CGFloat(100+20*i)
            
            bezierPath.move(to: CGPoint(x: x, y: 215))
            bezierPath.addLine(to: CGPoint(x: x, y: y))
        }
        barChart.path = bezierPath.cgPath
        barChart.strokeColor = UIColor.green.cgColor
        
        ani.keyPath = "strokeEnd"
        ani.timingFunction = CAMediaTimingFunction(name: .easeInEaseOut)
        ani.fromValue = 0.0
        ani.toValue = 1.0
        ani.autoreverses = false
        ani.duration = 2
        
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func barAniStart() {
        
        barChart.strokeEnd = 1.0
        barChart.add(ani, forKey: nil)
    }
    
}
